src/ucs/type/spinlock.h

```diff
@@ -40,50 +51,91 @@ typedef struct ucs_recursive_spinlock {
 } ucs_recursive_spinlock_t;
 
 
-static ucs_status_t ucs_spinlock_init(ucs_spinlock_t *lock, int flags)
+void ucs_recursive_spinlock_destroy(ucs_recursive_spinlock_t *lock);
+
+/**
+ * Spinlock implementation section
+ */
+static UCS_F_ALWAYS_INLINE void ucs_spinlock_init(ucs_spinlock_t *lock)
 {
-    int ret, lock_flags;
+    lock->lock = UCS_SPINLOCK_FREE;
+}
 
-    if (flags & UCS_SPINLOCK_FLAG_SHARED) {
-        lock_flags = PTHREAD_PROCESS_SHARED;
-    } else {
-        lock_flags = PTHREAD_PROCESS_PRIVATE;
+static UCS_F_ALWAYS_INLINE void ucs_spinlock_destroy(ucs_spinlock_t *lock)
+{
+}
+
+static UCS_F_ALWAYS_INLINE int ucs_spin_try_lock(ucs_spinlock_t *lock)
+{
+    int res;
+
+    res = ucs_atomic_cswap32(&lock->lock, UCS_SPINLOCK_FREE,
+                             UCS_SPINLOCK_BUSY) == UCS_SPINLOCK_FREE;
+
+#if defined(__powerpc64__)
+    /* barrier on locked spin */
+    if (res) {
+        asm volatile ("isync " ::: "memory");
     }
+#endif
 
-    ret = pthread_spin_init(&lock->lock, lock_flags);
-    if (ret != 0) {
-        return UCS_ERR_IO_ERROR;
+    return res;
+}
+
+static UCS_F_ALWAYS_INLINE void ucs_spin_lock(ucs_spinlock_t *lock)
+{
+    while (!ucs_spin_try_lock(lock)) {
+        while (lock->lock != UCS_SPINLOCK_FREE) {
+            /* spin */
+#if defined(__x86_64__)
+            asm volatile("pause" ::: "memory");
+#elif defined(__aarch64__)
+            asm volatile ("wfe"); /* suspend until event register is set */
+#elif defined(__powerpc64__)
+            asm volatile ("lwsync \n" \
+                          "isync  \n" \
+                          ::: "memory");
+#endif
+        }
     }
+}
 
-    return UCS_OK;
+static UCS_F_ALWAYS_INLINE void ucs_spin_unlock(ucs_spinlock_t *lock)
+{
+    static unsigned int lockfree = UCS_SPINLOCK_FREE;
+
+#if defined(__powerpc64__)
+    asm volatile("lwsync" : : : "memory");
+#endif
+
+    __atomic_store(&lock->lock, &lockfree, __ATOMIC_RELAXED);
+    
+#if defined(__aarch64__)
+    asm volatile ("sevl"); /* set event register */
+#endif
```

Comment:
there are too many differences now between architectures and code became "hairy"
maybe move it to new files: ucs/arch/xx/lock.h